### PR TITLE
Let marks be cleared by `clear` (and friends)

### LINKED ISSUE #15426: Scrollbar marks for shell integration do not clear when you use clear/cls commands
### Windows Terminal version

1.18.1421.0

### Windows build number

Version 10.0.22621.1702

### Other Software

PowerShell 7.3.4

### Steps to reproduce

1. Enable shell integration as documented here: https://learn.microsoft.com/en-us/windows/terminal/tutorials/shell-integration
2. Enable PWSH and CMD support (for PWSH I added it to my Documents\PowerShell\Profile.ps1)
3. Open new PWSH and CMD shells and press enter a bunch of times to generate a bunch of marks.
4. Type clear in PWSH or cls in CMD
6. Type commands that generate output like dir 
7. Press Ctrl+Up/Down a bunch

### Expected Behavior

The scrollbar marks should disappear on clear/cls and the Ctrl+Up/Down should navigate only through the visible commands

### Actual Behavior

The marks remain on the scrollbar and they will still be jumped to even though they are in the middle of command output.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Move scroll marks to `TextBuffer`, so they can be cleared by EraseInDisplay and EraseScrollback.

Also removes the namespacing on them.

## References and Relevant Issues
* see also #11000 and #15057
* Resize/Reflow _doesn't_ work yet and I'm not attempting this here. 

## Validation Steps Performed

* `cls` works
* `Clear-Host` works
* `clear` works
* the "Clear buffer" action works
* They work when there's marks above the current viewport, and clear the scrollback
* they work if you clear multiple "pages" of output, then scroll back to where marks previously were
* resizing doesn't totally destroy the marks

## PR Checklist
- [x] Closes #15426

### CODE REVIEW

File: src/buffer/out/textBuffer.cpp [hunk 1]
```diff
@@ -2869,3 +2872,66 @@ PointTree TextBuffer::GetPatterns(const til::CoordType firstRow, const til::Coor
     PointTree result(std::move(intervals));
     return result;
 }
+
+std::vector<ScrollMark>& TextBuffer::GetMarks()
+{
+    return _marks;
+}
```
Comment (lhecker) [inline src/buffer/out/textBuffer.cpp:2879]: I don't think this function is a good idea.

For instance, it leaks the way marks are stored. A `vector` is not a good way to store them and a ring buffer would be a better choice, because it makes the 2 most common operations much cheaper (adding at the end, removing at the beginning during buffer circling). Changing the underlying storage might require changing all callers around.
More importantly however I believe it's more bug-prone, because it allows the caller to make modifications that `TextBuffer` doesn't know about. Like, let's say we store marks sorted by start coordinate in the future to make range-based access and removal cheaper (`n` -> `log n`). If I change the `start` coordinate outside of `TextBuffer` then, this might break the sort order and cause the range-based functions to fail in weird ways. This might be obvious now, but I'm sure `COOKED_READ_DATA` was also obvious to someone at some point and now look at it (and us maintaining it). 😄 It has the exact same design flaw. It leaks its internals (and so does `GetRowByOffset` but I'm slowly working on fixing that).

File: src/buffer/out/textBuffer.hpp [hunk 2]
```diff
@@ -64,6 +64,41 @@ namespace Microsoft::Console::Render
     class Renderer;
 }
 
+enum class MarkCategory : size_t
```
Comment (lhecker) [inline src/buffer/out/textBuffer.hpp:67]: FYI you can use `uint8_t` here or just remove the explicit underlying type entirely.

File: src/cascadia/TerminalCore/TerminalApi.cpp [hunk 3]
```diff
@@ -458,29 +458,11 @@ void Terminal::NotifyBufferRotation(const int delta)
     // manually erase our pattern intervals since the locations have changed now
     _patternIntervalTree = {};
 
-    const auto hasScrollMarks = _scrollMarks.size() > 0;
+    auto& marks{ _activeBuffer().GetMarks() };
+    const auto hasScrollMarks = marks.size() > 0;
```
Comment (lhecker) [inline src/cascadia/TerminalCore/TerminalApi.cpp:462]: For instance, you can move this check into `ScrollMarks` itself.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 4]
```diff
@@ -3145,6 +3145,12 @@ bool AdaptDispatch::_EraseScrollback()
     auto& cursor = textBuffer.GetCursor();
     const auto row = cursor.GetPosition().y;
 
+    // Scroll all the marks up. This will trim ones that are now "outside" the buffer
+    textBuffer.ScrollMarks(-top);
+    // Also clear all the marks below the new viewport position.
+    textBuffer.ClearMarksInRange(til::point{ 0, height },
+                                 til::point{ bufferSize.width, bufferSize.height });
```
Comment (lhecker) [inline src/terminal/adapter/adaptDispatch.cpp:3152]: You could swap these two to reduce the amount of work. First clear everything outside of the `viewport` and then scroll the remaining bits up by `-top`.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 5]
```diff
@@ -2391,20 +2391,20 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+            [](const ::ScrollMark& m) { return til::point_span{ m.end, *m.commandEnd }; });
     }
     void ControlCore::ContextMenuSelectOutput()
     {
         _contextMenuSelectMark(
             _contextMenuBufferPosition,
-            [](const DispatchTypes::ScrollMark& m) -> bool { return !m.HasOutput(); },
-            [](const DispatchTypes::ScrollMark& m) { return til::point_span{ *m.commandEnd, *m.outputEnd }; });
+            [](const ::ScrollMark& m) -> bool { return !m.HasOutput(); },
+            [](const ::ScrollMark& m) { return til::point_span{ *m.commandEnd, *m.outputEnd }; });
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:2402]: You could use `auto` here.

**Review by lhecker — APPROVED** [review]
✅ because it works just like before, but...

If you don't mind, I think it would be better if you did add early-returns (`if (_marks.empty()) return;`) to all of the relevant functions. This makes it essentially cost free for `OpenConsole` and `conhost` (we could also use feature flags, but I'm personally impartial about that).

And additionally if the `GetMarks` function that returns a mutable reference could be removed. Reading through the changes I think this should be possible by adding:
* `AddMark(const ScrollMark& m, bool fromUi)` (I forgot, what does `fromUi` do?)
* `UpdatePromptEnd(til::point end)` (this can create the prompt mark if it doesn't already exist)
* `UpdateCommandEnd(til::point end)` (same here)
* ...

Basically, it would move more code from `Terminal` into `TextBuffer`, but that would IMO be very beneficial for long term maintenance once we aren't working here anymore. It would also make future improvements simpler, because it decouples the way marks are stored from their (internal) API.

**Review by DHowett — CHANGES_REQUESTED** [review]
I'm promoting Leonard's concerns into a block to signal agreement :smile:

File: src/buffer/out/textBuffer.cpp [hunk 6]
```diff
@@ -2924,6 +2920,18 @@ void TextBuffer::ScrollMarks(const int delta)
     _trimMarksOutsideBuffer();
 }
 
+void TextBuffer::AddMark(ScrollMark& m, const bool activeMark)
```
Comment (DHowett) [inline src/buffer/out/textBuffer.cpp]: API nit: `AddMark(..., true)` and `AddMark(..., false)` are unergonomic. What behavior do they capture? Why is `activeMark` pronounced "Active Mark" but defined as "insert mark at position zero?"

File: src/buffer/out/textBuffer.cpp [hunk 7]
```diff
@@ -2935,3 +2943,32 @@ void TextBuffer::_trimMarksOutsideBuffer()
                                 }),
                  _marks.end());
 }
+
+void TextBuffer::UpdateCurrentPromptEnd(const til::point pos)
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/buffer/out/textBuffer.cpp]: Or perhaps, `UpdateCurrentMarkPromptEnd`

Or, perhaps, this is actually a _setter_ whose name should be `SetCurrentMarkPromptEnd` since it takes in a position

File: src/buffer/out/textBuffer.cpp [hunk 8]
```diff
@@ -2935,3 +2943,32 @@ void TextBuffer::_trimMarksOutsideBuffer()
...
+void TextBuffer::UpdateCurrentCommandEnd(const til::point pos)
+{
+    if (_marks.empty())
+    {
+        return;
+    }
+    auto& curr{ _marks.back() };
+    curr.commandEnd = pos;
+}
+void TextBuffer::UpdateCurrentOutputEnd(const til::point pos, ::MarkCategory category)
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/buffer/out/textBuffer.cpp:2965]: aye

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 9]
```diff
@@ -1359,7 +1359,7 @@ void Terminal::AddMark(const ScrollMark& mark,
     m.end = end;
 
     // If the mark came from the user adding a mark via the UI, don't make it the active prompt mark.
-    _activeBuffer().AddMark(m, !fromUi);
+    fromUi ? _activeBuffer().AddMark(m) : _activeBuffer().StartPromptMark(m);
```
Comment (DHowett) [inline src/cascadia/TerminalCore/Terminal.cpp:1362]: It is very rare that using a ternary like this is the right thing to do.

File: src/buffer/out/textBuffer.cpp [hunk 10]
```diff
@@ -2920,16 +2920,13 @@ void TextBuffer::ScrollMarks(const int delta)
...
-        _marks.push_back(m);
-    }
-    else
-    {
-        _marks.insert(_marks.begin(), m);
-    }
+    _marks.push_back(m);
+}
+void TextBuffer::AddMark(const ScrollMark& m)
+{
```
Comment (DHowett) [inline src/buffer/out/textBuffer.cpp]: I still don't understand why "prompt mark" is "end of the entire list of marks" and "add mark" is the "beginning". I was hoping to find some clarity... but can you please just explain the guiding principle here?

Comment (DHowett) [reply src/buffer/out/textBuffer.cpp]: Reading other comments, I understand that "AddMark" is "add a mark that is not the active mark" and "StartPromptMark" is "add a mark that IS the active mark" :smile:

**Review by DHowett — APPROVED** [review]

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 9 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalCore/Terminal.cpp:1362]: yea that felt dirty to write
